### DISCUSSION TITLE
Challenges in Training YOLO Models with Large Images for Coral TPU Deployment

### DISCUSSION THREAD

**Discussion by heisenburrger** [ultralytics/ultralytics#22923] [Q&A]
I have a YOLO model trained on my custom dataset. I originally set the image size to 1792 because my dataset images have a resolution of 1512×2040. However, I need to run the model on a Google Coral TPU, which requires an input image size of 416.

When I train the model using an image size of 416, the accuracy becomes very poor. I think this might be because my dataset has a different aspect ratio, so training with 416×416 images may not be ideal. But since my dataset is already prepared and annotated at 1512×2040, I’m not sure what to do.

For inference, I know I can use tiling, but can I also use tiling for training? Do I need to split every image into 416×416 tiles along with their corresponding annotations?

What is the best approach here?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @heisenburrger, thanks for starting this detailed discussion and for using Ultralytics YOLO for your Coral TPU deployment work 🚀  

This is an automated response to help move things forward quickly — an Ultralytics engineer will also review your post and follow up with more specific guidance soon 🙂

In the meantime, to help us better understand your scenario (large images, aspect ratio differences, and Coral TPU 416-size requirement), please share as many of the following details as you can:

1. Which Ultralytics YOLO model you are using (e.g., YOLO11, YOLO26, etc.), and the exact `ultralytics` package version  
2. Your training configuration (image size, batch size, augmentation settings, `model` and `data` YAMLs if possible)  
3. A few representative example images from your dataset, including their annotations (especially where detections become poor when trained at 416)  
4. Training logs/plots (loss curves, mAP, precision/recall) comparing:  
   - Training at your original large image size  
   - Training at 416  
5. Any preprocessing or tiling steps you have already tried (for either training or inference)

We also recommend checking the core docs while you wait, as many similar topics around image sizes and training setups are covered there:

- General docs: [Docs](https://docs.ultralytics.com/)  
- Python examples: [Python](https://docs.ultralytics.com/usage/python/)  
- CLI usage: [CLI](https://docs.ultralytics.com/usage/cli/)  

If this discussion eventually reveals a training-related issue or limitation, it can be very helpful if you can narrow things down as much as possible to a small, clearly described configuration (for example, a subset of your data and a minimal training command). That makes it much easier for us to reason about behavior and reproduce any potential problems.

If at any point this turns into a 🐛 Bug Report (for example, if you see inconsistent behavior between image sizes or environments), please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) so our team can investigate efficiently.

If this remains a custom training ❓ Question (which it sounds like for now), please continue to:

- Include dataset image examples (especially showing your 1512×2040 samples and any tiling you are considering)  
- Share your training logs and metrics  
- Verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/), as these practices often impact performance when changing image sizes or deployment targets like Coral TPU

You’re also very welcome to discuss tiling strategies and deployment considerations with the community:

- Real-time chat: [Discord](https://discord.com/invite/ultralytics) 🎧  
- In-depth discussions: [Discourse](https://community.ultralytics.com/)  
- Community threads and experiences: [Subreddit](https://www.reddit.com/r/Ultralytics/)  

## Upgrade

Before we dive deeper, please ensure you are on the latest `ultralytics` version (many training and export improvements land there), and that your environment is up to date:

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a **Python>=3.8** environment with **PyTorch>=1.8** to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including CUDA/CUDNN, Python and PyTorch preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

Using one of these can help rule out environment-specific issues when comparing training at large image sizes versus 416 for Coral TPU deployment.

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Reply by glenn-jocher** [reply]
You’re diagnosing this correctly: when you drop from 1512×2040 to 416 input, YOLO is mainly losing raw resolution (small objects shrink a lot), not being “confused” by aspect ratio, since training automatically resizes while preserving aspect ratio via `imgsz` and letterboxing.

If you need Coral’s fixed 416 input but want something closer to your 1792‑imgsz performance, the usual approach is exactly what you suggested: tile the large images offline into 416×416 crops (often with some overlap), remap the box coordinates into each crop, and then train directly on those tiles with `imgsz=416` so training and inference see the same effective scale. This is the same idea described as “image tiling for small objects” in the [model evaluation and fine‑tuning guide](https://docs.ultralytics.com/guides/model-evaluation-insights/), and the general workflow of splitting images and updating labels is outlined in the [preprocessing annotated data guide](https://docs.ultralytics.com/guides/preprocessing_annotated_data/).

At inference time on Coral, you then either 1) run the same 416×416 tiles through the exported Edge TPU model (exported with `imgsz=416` as shown in the [Edge TPU integration docs](https://docs.ultralytics.com/integrations/edge-tpu/)), or 2) keep full‑image workflow but use sliced inference on the host (for example, following the [SAHI tiled inference guide](https://docs.ultralytics.com/guides/sahi-tiled-inference/)) and send each slice to the Coral.
